File: chat_server.py
# chat_server.py - Сервер локального чата на Python (сокеты, threading)
import socket
import threading
import json
import time

class ChatServer:
    def __init__(self, host='127.0.0.1', port=5555):
        self.server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.server.bind((host, port))
        self.server.listen()
        self.clients = []
        self.nicknames = []
        print(f"[SERVER] Запущен на {host}:{port}")

    def broadcast(self, message, sender_socket=None):
        for client in self.clients:
            if client != sender_socket:
                try:
                    client.send(message.encode('utf-8'))
                except:
                    self.remove_client(client)

    def remove_client(self, client):
        if client in self.clients:
            idx = self.clients.index(client)
            self.clients.remove(client)
            nickname = self.nicknames.pop(idx)
            self.broadcast(f"[SERVER] {nickname} покинул чат")
            print(f"[SERVER] {nickname} отключился")

    def handle_client(self, client):
        while True:
            try:
                msg = client.recv(1024).decode('utf-8')
                if not msg:
                    break
                data = json.loads(msg)
                if data['type'] == 'msg':
                    formatted = f"[{data['nickname']}]: {data['text']}"
                    print(formatted)
                    self.broadcast(formatted, client)
                elif data['type'] == 'private':
                    target_nick = data['target']
                    # найти клиента по нику
                    for i, nick in enumerate(self.nicknames):
                        if nick == target_nick:
                            self.clients[i].send(json.dumps({'type':'private','from':data['nickname'],'text':data['text']}).encode())
                            break
                elif data['type'] == 'file':
                    # отправка файла (base64)
                    self.broadcast(f"[FILE] {data['nickname']} отправил файл: {data['filename']}", client)
                    for i, nick in enumerate(self.nicknames):
                        if nick == data['nickname']: continue
                        self.clients[i].send(json.dumps({'type':'file','from':data['nickname'],'filename':data['filename'],'data':data['data']}).encode())
            except:
                self.remove_client(client)
                break

    def start(self):
        while True:
            client, addr = self.server.accept()
            print(f"[SERVER] Подключен {addr}")
            client.send("NICK".encode('utf-8'))
            nickname = client.recv(1024).decode('utf-8')
            self.nicknames.append(nickname)
            self.clients.append(client)
            self.broadcast(f"[SERVER] {nickname} присоединился к чату")
            thread = threading.Thread(target=self.handle_client, args=(client,))
            thread.start()

if __name__ == "__main__":
    server = ChatServer()
    server.start()

File: ChatServer.java
// ChatServer.java - Сервер чата на Java (ServerSocket)
import java.io.*;
import java.net.*;
import java.util.*;

public class ChatServer {
    private static Set<ClientHandler> clients = new HashSet<>();
    public static void main(String[] args) throws IOException {
        ServerSocket serverSocket = new ServerSocket(5555);
        System.out.println("Сервер запущен на порту 5555");
        while (true) {
            Socket socket = serverSocket.accept();
            ClientHandler handler = new ClientHandler(socket);
            clients.add(handler);
            handler.start();
        }
    }
    static void broadcast(String message, ClientHandler exclude) {
        for (ClientHandler client : clients) {
            if (client != exclude) {
                client.sendMessage(message);
            }
        }
    }
    static void removeClient(ClientHandler client) {
        clients.remove(client);
        broadcast("SERVER: " + client.getNickname() + " покинул чат", null);
    }
}

class ClientHandler extends Thread {
    private Socket socket;
    private BufferedReader in;
    private PrintWriter out;
    private String nickname;
    public ClientHandler(Socket socket) {
        this.socket = socket;
    }
    public void run() {
        try {
            in = new BufferedReader(new InputStreamReader(socket.getInputStream()));
            out = new PrintWriter(socket.getOutputStream(), true);
            out.println("NICK");
            nickname = in.readLine();
            System.out.println(nickname + " подключился");
            ChatServer.broadcast("SERVER: " + nickname + " присоединился к чату", this);
            String inputLine;
            while ((inputLine = in.readLine()) != null) {
                if (inputLine.startsWith("/w ")) {
                    String[] parts = inputLine.split(" ", 3);
                    if (parts.length >= 3) {
                        String target = parts[1];
                        String msg = parts[2];
                        sendPrivate(target, msg);
                    } else {
                        out.println("SYSTEM: Используйте /w ник сообщение");
                    }
                } else {
                    ChatServer.broadcast(nickname + ": " + inputLine, this);
                }
            }
        } catch (IOException e) {
            System.out.println("Ошибка: " + e.getMessage());
        } finally {
            try { socket.close(); } catch (IOException e) {}
            ChatServer.removeClient(this);
        }
    }
    private void sendPrivate(String target, String msg) {
        for (ClientHandler client : ChatServer.clients) {
            if (client.getNickname().equals(target)) {
                client.sendMessage("[Приват от " + nickname + "]: " + msg);
                this.sendMessage("[Приват для " + target + "]: " + msg);
                return;
            }
        }
        this.sendMessage("SYSTEM: Пользователь " + target + " не найден");
    }
    public void sendMessage(String msg) { out.println(msg); }
    public String getNickname() { return nickname; }
}
